stats/utils/src/main/java/org/apache/bookkeeper/stats/utils/StatsDocGenerator.java

```diff
@@ -0,0 +1,302 @@
+        if (Strings.isNullOrEmpty(file)) {
+            writer = new StringWriter();
+        } else {
+            writer = new OutputStreamWriter(new FileOutputStream(file), UTF_8);
+        }
+        try {
+            Map<String, Map<String, Map<String, String>>> docs = docEntries.entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                    e -> e.getKey(),
+                    e -> e.getValue().entrySet()
+                        .stream()
+                        .collect(Collectors.toMap(
+                            e1 -> e1.getKey(),
+                            e1 -> e1.getValue().properties()
+                        ))
+                ));
+            yaml.dump(docs, writer);
+            writer.flush();
+            if (writer instanceof StringWriter) {
+                StringWriter sw = (StringWriter) writer;
+                System.out.println(sw.toString());
+            }
+        } finally {
+            writer.close();
+        }
+    }
+
+    /**
+     * Args for stats generator.
+     */
+    private static class MainArgs {
+
+        @Parameter(
+            names = {
+                "-p", "--package"
+            },
+            description = "Package prefix of the classes to generate stats doc")
+        String packagePrefix = "org.apache.bookkeeper";
+
+        @Parameter(
+            names = {
+                "-sp", "--stats-provider"
+            },
+            description = "The stats provider used for generating stats doc")
+        String statsProviderClass = "prometheus";
+
+        @Parameter(
+            names = {
+                "-o", "--output-file"
+            },
+            description = "The output file to dump stats docs"
+        )
+        String yamlFile = null;
+
+        @Parameter(
+            names = {
+                "-h", "--help"
+            },
+            description = "Show this help message")
+        boolean help = false;
+
+    }
+
+    public static void main(String[] args) throws Exception {
+        MainArgs mainArgs = new MainArgs();
+
+        JCommander commander = new JCommander();
+        try {
+            commander.setProgramName("stats-doc-gen");
+            commander.addObject(mainArgs);
+            commander.parse(args);
+            if (mainArgs.help) {
+                commander.usage();
+                Runtime.getRuntime().exit(0);
+                return;
+            }
+        } catch (Exception e) {
+            commander.usage();
+            Runtime.getRuntime().exit(-1);
+            return;
+        }
+
+        Stats.loadStatsProvider(getStatsProviderClass(mainArgs.statsProviderClass));
+        StatsProvider provider = Stats.get();
+
+        StatsDocGenerator docGen = new StatsDocGenerator(
+            mainArgs.packagePrefix,
+            provider
+        );
+        docGen.generate(mainArgs.yamlFile);
+    }
+
+    private static String getStatsProviderClass(String providerClass) {
```

Comment:
Presumably, this is going to be called from another script as part of the website/docs build chain. As such, if we add a new provider, we need to add it to that script and also add it here, which people will forget to do.

It would be better to scan the classpath for all implementors of StatsProvider, or at least add an option to do so for the website build use case.

Reply:
`dev/stats-doc-gen` is using bookkeeper-dist/all as its classpath, so it already scans all dependencies.

Reply:
How is this going to be called? And what will the output look like?

Reply:
Actually, I see the output in the description. What is the point of having a stats provider argument to the tool? if it is because each provider produces a different output and we need to have each output in our documentation, then we should have an option to not provide a provider, and have it output for all providers.

Reply:
yes. different provider will produce different output. this PR only provides tool but not yet hook up with website build. there is no point for providing this option in this PR for now. I would rather defer this until we want to hook up with website building.

Reply:
yes. as the title says, this PR only introduces the annotation and a tool. we have to put all annotations to metrics before we do the actual website change. I can't do it all in one PR.